### PR TITLE
Bundled multidex jar

### PR DESCRIPTION
**Pull request by atsushieno**
We need to switch multidex support library from obsolete support package to m2repository. (Because current one in use is, obsoleted. Those who set custom Android SDK cannot find it because they don't know much about Android SDK...)

Switching to m2repository is not that simple. There is no android-support-multidex.jar anymore. So far I can think of two solution options described below:

1: bundle android-support-multidex.jar in the resulting SDK just as a binary. It is the safest, most reliable and almost backward compatible solution. We'll just have to copy the jar to bin/Debug/lib/mandroid.
  - Optionally, extract jar from m2repository instead of having binary checkin (which is almost harmless because it's just 21kb, but still needs some MSBuild targets hack that is messy anyways).

2: implement all below:
  - Add m2repository download to android-sdk-tool and xamarin-android/build-tools/android-toolchains. It is not very complicated.
  - Make changes to Xamarin.Android.Common.targets in MSBuild tasks in xamarin-android to do either of below (note that both likely need to be implemented in a new Task):
    - extract the jar from the relevant aar, cache it, and add it as AndroidJavaLibrary as it used to do.
    - implement additional build item support that we can retrieve only part of an aar, like we have in Binding library support.
  - Implement additional clean step to ensure that the jar is removed.
  - Add tests to ensure that ALL ABOVE works.

I believe there is no reason to chose option 2 in terms of implementation/QA cost. Just take the easier option.

### CODE REVIEW

**Comment by jonpryor** [comment]
> 1. bundle android-support-multidex.jar in the resulting SDK just as a binary.

That sounds like we should distribute this file within the Xamarin.Android SDK. *Can we do that?* What is the redistributable license for it?

Option 2 is more complicated, but it can be done on the end-users machine, and thus it doesn't have any redistribution requirements like Option 1 does.

**Comment by atsushieno** [comment]
It should be APACHE2 as descibed in android-sdk-*/extras/android/m2repository/NOTICE.txt.

**Comment by atsushieno** [comment]
Maybe we should copy that as something like lib/mandroid/MULTIDEX_JAR_LICENSE.txt to make it explicit.

**Comment by jonpryor** [comment]
If we copy the license and update e.g. `/Library/Frameworks/Xamarin.Android.framework/Versions/Current/License.rtf`/etc. to also mention this file -- I believe we have guidelines for this somewhere -- then bundling and redistributing it should be fine.

**Comment by atsushieno** [comment]
Where can I find the *existing* license file? There must be one since we have already including a lot of third party softwares this must not be the first one, right?

**Comment by atsushieno** [comment]
A better solution is, as I initially suggested as an example solution, copy the support library license to lib/mandroid as MULTIDEX_JAR_LICENSE because just having one License.rtf in the product does not resolve the problem that people cannot find the exact relevant license for the .jar file.

**Comment by azchohfi** [comment]
@atsushieno The same thing (using m2repository) should be done for **proguard.jar**, since it also uses the deprecated one (android-sdk\tools\proguard\lib\proguard.jar). That version is still ProGuard 4.7, which still does not support JDK1.8.

**Comment by atsushieno** [comment]
It is not exactly the same code path so I won't handle it within *this* specific PR, but could you tell me in which directory in Android support m2repository it should be? I cannot locate it.

**Comment by atsushieno** [comment]
My current plan to support Java8 class file is to handle it via jack, which seems to also handle proguard stuff. That needs not a small set of changes (but not very big).

**Comment by azchohfi** [comment]
It is not, my mistake. Gradle downloads it, but it's not inside m2repo.
I've manually downloaded it from here:
https://sourceforge.net/projects/proguard/files/proguard/5.2/proguard5.2.1.zip/download
Replacing proguard with version 5.2.1 makes it works with 1.8, but yeah, that should be handled in a different PR, and even better if it will be handled using Jack (awesome by the way)!

**Comment by atsushieno** [comment]
Oh okay, that actually still makes it much easier. Thanks for the great input!

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets [hunk 1]
```diff
@@ -1,9 +1,18 @@
...
   <UsingTask TaskName="Xamarin.Android.Tools.BootstrapTasks.GenerateProfile" AssemblyFile="..\..\bin\Build$(Configuration)\Xamarin.Android.Tools.BootstrapTasks.dll" />
+  <UsingTask TaskName="Xamarin.Android.Tools.BootstrapTasks.UnzipDirectoryChildren" AssemblyFile="..\..\bin\Build$(Configuration)\Xamarin.Android.Tools.BootstrapTasks.dll" />
   <PropertyGroup>
     <_SharedRuntimeBuildPath Condition=" '$(_SharedRuntimeBuildPath)' == '' ">..\..\bin\$(Configuration)\lib\xbuild-frameworks\MonoAndroid\</_SharedRuntimeBuildPath>
     <_GeneratedProfileClass>$(IntermediateOutputPath)Profile.g.cs</_GeneratedProfileClass>
+    <BuildDependsOn>
+		$(BuildDependsOn);
+		_CopyExtractedMultiDexJar;
+    </BuildDependsOn>
+    <_AndroidSdkLocation>$(ANDROID_SDK_PATH)</_AndroidSdkLocation>
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: This should *probably* be the [`$(AndroidSdkFullPath )` MSBuild property](https://github.com/xamarin/xamarin-android/blob/4fd61bc/Configuration.props#L59). Nothing is going to export an `$(ANDROID_SDK_PATH)` environment variable.

Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: Unrelatedly: Please fix your editor to expand tabs into spaces on `.targets` files and other XML files, with 2 space indents.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets [hunk 2]
```diff
@@ -19,4 +28,19 @@
         Lines="$(_GeneratedProfileClass)"
         Overwrite="false"/>
   </Target>
+  
+  <Target Name="_CopyExtractedMultiDexJar"
+    Inputs="$(_AndroidSdkLocation)\$(_MultiDexAarInAndroidSdk);$(_AndroidSdkLocation)\$(_SupportLicenseInAndroidSdk)"
+    Outputs="$(OutputPath)..\..\..\mandroid\android-support-multidex.jar;$(OutputPath)..\..\..\mandroid\MULTIDEX_JAR_LICENSE">
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: Should `MULTIDEX_JAR_LICENSE` have a `.txt` extension?

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets (hunk 1 above)
Comment (atsushieno) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: No. monodroid does export that environment variable. Want me to eliminate the variables that are used ONLY in monodroid in the same manner? For example, InstantRunEnabled.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets (hunk 2 above)
Comment (atsushieno) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: Too cosmetic to change.

File: src/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/UnzipDirectoryChildren.cs [hunk 3]
```diff
@@ -29,6 +29,8 @@ public UnzipDirectoryChildren ()
 		[Required]
 		public  ITaskItem       DestinationFolder   { get; set; }
 
+		public bool NoSubdirectory { get; set; }
```
Comment (jonpryor) [inline src/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/UnzipDirectoryChildren.cs]: I don't like this property name. Let's try to find a better one...

The reason the task is `UnzipDirectoryChildren` and not `Unzip` is precisely *because* it "skips"/"ignores" the topmost extracted directory, hence the `Directory.EnumerateDirectories(nestedTemp, "*")`.

In an ideal world, I think we'd want a separate task, but that's also ugly.

Perhaps `IncludeRootDirectory` would be clearer?

I'll I'm sure of is that `NoSubdirectory` doesn't seem clear.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets (hunk 1 above)
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: What should happen for "pure" xamarin-android builds? `$(ANDROID_SDK_PATH)` won't be set, so e.g. `$(_AndroidSdkLocation)\$(_MultiDexAarInAndroidSdk)` will be `/extras/android/...`, which won't be readable or writable, and might result in a xamarin-android build error.

That seems less than desirable.

Perhaps this should use a fallback?

    <_AndroidSdkLocation Condition=" '$(AndroidSdkFullPath)' != '' ">$(AndroidSdkFullPath)</_AndroidSdkLocation>
    <_AndroidSdkLocation Condition=" '$(_AndroidSdkLocation)' == '' ">$(ANDROID_SDK_PATH)</_AndroidSdkLocation>

This should allow xamarin-android to build while allowing the monodroid build to override.

Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: > Want me to eliminate the variables that are used ONLY in monodroid in the same manner? For example, InstantRunEnabled.

If they *can* be removed, that would likely be best. If they *can't* be removed -- which may be the case for `$(_InstantRunEnabled)`, which is used with the `<DetermineJavaLibrariesToCompile/>` task -- then perhaps it should be renamed to be something more meaningful w/o context/memory/history.

For example, instead of `DetermineJavaLibrariesToCompile.EnableInstantRun`, rename it to `DetermineJavaLibrariesToCompile.SkipMonoPlatformJar`. For those just reading `Xamarin.Android.Common.targets`, I assume that seeing `SkipMonoPlatformJar="$(_SkipMonoPlatformJar)"` would be more meaningful than `EnableInstantRun="$(_InstantRunEnabled)"`.

File: src/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/UnzipDirectoryChildren.cs (hunk 3 above)
Comment (atsushieno) [inline src/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/UnzipDirectoryChildren.cs]: This change follows the "keep changes minimum" principle. If you make your suggested change, that affects other build targets. Such a semantic change should be done in separate PR.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets (hunk 1 above)
Comment (atsushieno) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets]: The pure xamarin-android configuration for _AndroidSdkLocation is defined immediately below.
